### DISCUSSION TITLE
Updating MIDI data in project?

### DISCUSSION THREAD

**Discussion by xhevahir** [demberto/PyFLP#105] [General]
Hi, I'm new to FL Studio, so I don't know the terminology well enough to figure out from the docs whether it's possible to get the MIDI recordings contained in a project and substitute them with new MIDI files.

It looks like MIDI functionality in PyFLP is under development. Does an FLP even contain references to MIDI files, or only an internal representation of their data?

Thanks for any help. I've been trying to solve this problem through automating the UI (which looks difficult) and wondered if your project might provide a better way.

**Comment by demberto** [comment]
The FLP format uses a slightly different format for storing pattern notes (as defined in `pyflp/pattern.py` in the `NotesEvent` class). No MIDI files (or references to them) are stored.

Taking the title of the discussion literally, its possible to edit the existing note data with PyFLP but I have yet to add a function that lets you add / delete notes.

I am working on a FLP <-> MIDI exporter rn, but it seems that time signatures make offset calculations harder. Work on it has been slow as `mido` seems to be the only good library for parsing and creating MIDI files and its untyped and unmaintained. However I think it will work correctly for 4/4 patterns.

I highly recommend you to read the FL Studio manual once to get some terminologies straight, since PyFLP's higher level API is designed to mimic FL Studio's GUI (as such you'd see "Channel Rack", "Arrangements" in the docs). I'd also like your opinion and suggestions about the documentation itself.
